src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcherState.scala

```diff
@@ -100,15 +106,19 @@ case class BlockFetcherState(
         }
       )
 
-  def addBodies(peer: Peer, bodies: Seq[BlockBody]): BlockFetcherState = {
-    val (matching, waiting) = waitingHeaders.splitAt(bodies.length)
-    val blocks = matching.zip(bodies).map((Block.apply _).tupled)
-
-    withPeerForBlocks(peer.id, blocks.map(_.header.number))
-      .copy(
-        readyBlocks = readyBlocks.enqueue(blocks),
-        waitingHeaders = waiting
-      )
+  def addBodies(peerId: PeerId, bodies: Seq[BlockBody]): BlockFetcherState = {
+    if (bodies.isEmpty) {
+      copy(waitingHeaders = Queue.empty)
```

Comment:
why do the clean up and not return just `this`?

Reply:
I agree with mk, this is inconsistent with the fact that we are keeping the `waitingHeaders = waiting` below

Reply:
How are we going to invalidate headers from rejected chain then? I thought that this is the way we agreed on to solve that scenario: https://github.com/input-output-hk/mantis/pull/787#issuecomment-726177268.

I don't think this is really an inconsistency since they are different cases (no bodies means either given node is not synced up or it simply doesn't have given blocks for any reason).

I can make it clearer by adding comment or turning the return type into an either (plus comment), but I do think that this is the way to go (and works really well as tests from last weekend show)

Reply:
Fair enough! i think just adding a comment like *this is enough:
*"no bodies means either given node is not synced up or it simply doesn't have given blocks for any reason. Faced with this scenario we decide to give up about those requested headers"

Reply:
Done exactly that thing ;)

Reply:
Agree as well! Had a whole chat with mk where he convinced me of this being the approach (I expected something like this but implemented more closer to what we have on the other project, though this approach fits better with having requests in parallel)